arbiter-core/src/tests/interaction.rs

```diff
@@ -101,121 +121,106 @@ async fn filter_watcher() -> Result<()> {
+            format!("new_{}", TEST_ARG_SYMBOL),
+            TEST_ARG_DECIMALS,
+        ),
+    )?
+    .send()
+    .await?;
+
+    // Sanity check that tokens have different addresses
+    assert_ne!(arbiter_token.address(), arbiter_token2.address());
+
+    let approval = arbiter_token2.approve(
+        client.default_sender().unwrap(),
+        ethers::types::U256::from(TEST_APPROVAL_AMOUNT),
+    );
+    approval.send().await?.await?;
 
-//     // Create a new token contract to check that the address watcher only
-// gets     // events from the correct contract Check that only the default
-// watcher gets     // this event
-//     let arbiter_token2 = ArbiterToken::deploy(
-//         client.clone(),
-//         (
-//             format!("new_{}", TEST_ARG_NAME),
-//             format!("new_{}", TEST_ARG_SYMBOL),
-//             TEST_ARG_DECIMALS,
-//         ),
-//     )?
-//     .send()
-//     .await?;
-//     let mint2 = arbiter_token2.mint(
-//         ethers::types::H160::from_str(TEST_MINT_TO)?,
-//         ethers::types::U256::from(TEST_MINT_AMOUNT),
-//     );
-//     mint2.send().await?.await?;
-//     let default_watcher_event = default_watcher.next().await.unwrap();
-//     assert!(!default_watcher_event.data.is_empty());
-//     println!("default_watcher_event: {:#?}", default_watcher_event);
+    // get the next event with the default_watcher
+    let event_two = default_watcher.next().await.unwrap();
+    assert!(!event_two.data.is_empty());
 
-//     // Use tokio::time::timeout to await the approval_watcher for a specific
-//     // duration The timeout is needed because the approval_watcher is a
-// stream     // that will never end when the test is passing
-//     let timeout_duration = tokio::time::Duration::from_secs(1); // Adjust the
-// duration as needed     let timeout = tokio::time::timeout(timeout_duration,
-// address_watcher.next());     match timeout.await {
-//         Result::Ok(Some(_)) => {
-//             // Event received
-//             panic!("This means the test is failing! The filter did not
-// work.");         }
-//         Result::Ok(None) => {
-//             // Timeout occurred, no event received
-//             println!("Expected result. The filter worked.")
-//         }
-//         Err(_) => {
-//             // Timer error (shouldn't happen in normal conditions)
-//             panic!("Timer error!")
-//         }
-//     }
-//     Ok(())
-// }
+    // check that the address_watcher has not received any events
+    let mut ctx = Context::from_waker(futures::task::noop_waker_ref());
+    let poll_result = Pin::new(&mut address_watcher).poll_next(&mut ctx);
+    match poll_result {
+        Poll::Ready(Some(_event)) => panic!("Event received unexpectedly!"),
+        Poll::Ready(None) => println!("Stream completed!"),
+        Poll::Pending => println!("No event ready yet, as expected."),
+    }
+    assert_matches!(poll_result, Poll::Pending);
+    Ok(())
```

Comment:
Okay this is the core idea that lets this all work. Can you explain it to me?

Reply:
yeah so the call to `.next().await()` will not normally terminating when in the `Poll::Pending` state so what i did was bring the poll result into scope and match on the poll result so that we can terminate the tests. 
- To poll a stream manually, you need a context (Context). The context, among other things, contains a "waker", which is an object that allows asynchronous tasks to notify when they're ready to make progress.
- A "noop waker" is a waker that does nothing when woken up. Since we're only interested in the current state of the stream and don't need to actually schedule any async tasks, a noop waker is sufficient.

The subsequent match statement handles each potential result from the poll_next:
- If an event is ready, which wasn't expected in this scenario, it panics.
- If the stream is complete, it prints that.
- If there's no event ready yet (i.e., it's still waiting for an event), it prints that.

By manually polling the stream, we are able to gain direct insight into the state of the address_watcher. Instead of using .await and blocking the task indefinitely.

Reply:
Nice! I think that makes sense to me. The context part is the most fuzzy though.